### PR TITLE
[Spritelab] Add Mini Toolbox to event blocks, behind experiment

### PR DESCRIPTION
**Pull request by ajpal**
Experiment flag: `miniToolbox`

Done:
- miniToolboxBlocks as a new option in the block config
- generated code changes to make everything work
- subject/object pointer blocks

To Do:
- set up shadowing for subject and object blocks
- make the blocks within the mini toolbox shadow the costume
- add the word "sprite" on the block when there's no costume to show
- Make it configurable on levelbuilder whether or not to show the minitoolbox

image

image

Without experiment:
image

### CODE REVIEW

File: apps/src/block_utils.js [hunk 1]
```diff
@@ -996,6 +1000,32 @@ exports.createJsWrapperBlockCreator = function(
...
+          experiments.isEnabled(experiments.MINI_TOOLBOX)
+        ) {
+          var toggle = new Blockly.FieldIcon('+');
+          var miniToolboxXml = '<xml>';
+          miniToolboxBlocks.forEach(block => {
+            miniToolboxXml += `\n <block type="${block}"></block>`;
+          });
+          miniToolboxXml += '\n</xml>';
+          this.tray = false;
+          Blockly.bindEvent_(toggle.fieldGroup_, 'mousedown', this, () => {
```
Comment (jmkulwik) [inline apps/src/block_utils.js:1016]: NIT: Mind leaving a quick comment explaining what's going on here? It makes sense next to the visual, but is less obvious when reading the code directly.

File: apps/src/p5lab/spritelab/blocks.js [hunk 2]
```diff
@@ -199,7 +199,18 @@ const customInputTypes = {
         .appendTitle(new Blockly.FieldImage('', 32, 32), inputConfig.name);
     },
     generateCode(block, arg) {
-      return `'${block.getTitleValue(arg.name)}'`;
+      switch (block.type) {
+        case 'gamelab_clickedSpritePointer':
+          return '{id: extraArgs.sprite}';
+        case 'gamelab_subjectSpritePointer':
+          return '{id: extraArgs.sprite}';
```
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/blocks.js:206]: For PR documentation: When I first read this, I was expecting to see three different IDs (i.e. `extraArgs.sprite` is here twice). Why is the `.sprite` ID reused?

**Review by jmkulwik — APPROVED** [review]
Super exciting that this is happening! I had one quick documentation ask, if you don't mind (and a nit). Besides that, good to go :)

File: apps/src/p5lab/spritelab/blocks.js (hunk 2 above)
Comment (ajpal) [reply apps/src/p5lab/spritelab/blocks.js:206]: When we detect an event, we call back into the interpreted code, passing in any relevant info in the `extraArgs` parameter. For click events, there's just one sprite, so the shape of that object is `{sprite: <some id>}`. For touch events, there are two sprites, so we have `{sprite: <some id>, target: <some id>}`
The generated code for the clicked/subject/object pointer blocks just need to evaluate to the appropriate sprite, referenced by its id, so the generated code should be `{id: <some id>}`. In the case of click events, the relevant id is passed in as `extraArgs.sprite` and in the case of touch events, the ids are `extraArgs.sprite` and `extraArgs.target`

**Pull request merged**